tensor/iterator_mult.go

```diff
@@ -0,0 +1,295 @@
+package tensor
+
+// MultIterator is an iterator that iterates over multiple tensors, including masked tensors.
+//  It utilizes the *AP of a Tensor to determine what the next index is.
+// This data structure is similar to Numpy's flatiter, with some standard Go based restrictions of course
+// (such as, not allowing negative indices)
+
+type MultIterator struct {
+	fit0      *FlatIterator
+	trackIdx  []int
+	fitArr    []*FlatIterator
+	lastIndex []*int
+	masks     [][]bool
+	termInt   int // Should be -1
+}
+
+func genIterator(m map[int]*FlatIterator, shape []int, strides []int) (*FlatIterator, int) {
+	key := hashIntArrayPair(shape, strides)
+	f := m[key]
+	if f == nil {
+		f = NewFlatIterator(&AP{shape: shape, strides: strides})
+		m[key] = f
+	}
+	return f, key
+}
+
+// NewMultIterator creates a new MultIterator from a list of APs
+func NewMultIterator(aps ...*AP) *MultIterator {
+	nit := len(aps)
+	if nit < 1 {
+		return nil
+	}
+
+	it := new(MultIterator)
+
+	if nit == 1 && !aps[0].IsMasked() {
+		it.fit0 = NewFlatIterator(&AP{shape: aps[0].shape, strides: aps[0].strides})
+		return it
+	}
+
+	for i := 1; i < len(aps); i++ {
+		if aps[i] != nil {
+			if !(aps[i].IsScalar()) {
+				for j := 0; j < len(aps); j++ {
+					if aps[j] != nil {
+						if !(aps[j].IsScalar()) {
+							if !EqInts(aps[i].shape, aps[j].shape) {
+								//panic("Can not iterate through tensors of different shape")
+								return nil
+							}
+						}
+					}
+				}
+				break
+			}
+		}
+	}
+
+	it.fitArr = make([]*FlatIterator, 2*nit)
+	it.lastIndex = make([]*int, 2*nit)
+	it.trackIdx = BorrowInts(2 * nit)
+	it.termInt = -1
+
+	m := make(map[int]*FlatIterator)
+
+	for i, ap := range aps {
+		it.lastIndex[2*i] = &it.termInt
+		it.lastIndex[2*i+1] = &it.termInt
+
+		if ap != nil {
+			if ap.IsMasked() {
+				f, key := genIterator(m, ap.shape, ap.maskStrides)
+				it.lastIndex[2*i+1] = &(f.lastIndex)
+				it.trackIdx[2*i+1] = key
+			}
+			f, key := genIterator(m, ap.shape, ap.strides)
+			it.lastIndex[2*i] = &(f.lastIndex)
+			it.trackIdx[2*i] = key
+		}
+	}
+
+	i := 0
+	for key, f := range m {
+		it.fitArr[i] = f
+		i++
+		for j := range it.trackIdx {
+			if it.trackIdx[j] == key {
+				it.trackIdx[j] = i
+			}
+		}
+	}
+	it.fitArr = it.fitArr[:i]
+	return it
+}
+
+// MultIteratorFromDense creates a new MultIterator from a list of dense tensors
+func MultIteratorFromDense(tts ...*Dense) *MultIterator {
+	aps := BorrowAPList(len(tts))
+	var masked = false
+	for i, tt := range tts {
+		aps[i] = tt.Info()
+		masked = masked || aps[i].IsMasked()
+	}
+	it := NewMultIterator(aps...)
+	if masked {
+		masks := BorrowMaskList(len(tts))
+		for i, tt := range tts {
+			masks[i] = tt.mask
+		}
+		it.masks = masks
+	}
+	ReturnAPList(aps)
+	return it
+}
+
+// DestroyMultIterator creates a new MultIterator from a list of dense tensors
```

Comment:
maybe unexport this,

Then in any function that creates `MultIterator` add `runtime.SetFinalizer` to call this function?